### ISSUE TITLE
Emit ResourceWarning when GzipFile is deleted with unwritten data

### ISSUE THREAD

**Issue by cmaloney** [python/cpython#130806]
# Feature or enhancement

### Proposal:

This may indicate accidental data loss.

Ways to make sure all data is written:
1. Use the [file-like object](https://docs.python.org/3/glossary.html#term-file-object) as a [“With Statement Context Manager”](https://docs.python.org/3/reference/datamodel.html#context-managers). 
   - All objects which [inherit](https://docs.python.org/3/tutorial/classes.html#inheritance) from [IOBase](https://docs.python.org/3/library/io.html#io.IOBase) support this.
   - [`BufferedIOBase`](https://docs.python.org/3/library/io.html#io.BufferedIOBase), [`BufferedWriter`](https://docs.python.org/3/library/io.html#io.BufferedWriter), and [`GzipFile`](https://docs.python.org/3/library/gzip.html#gzip.GzipFile) all support this.
   - Ensures `.close()` is called in both exception and regular cases.
1. Ensure [`.close()`](https://docs.python.org/3/library/io.html#io.IOBase.close) is always called which flushes data before closing.
1. If the underlying stream need to be kept open, use [`.detach()`](https://docs.python.org/3/library/io.html#io.BufferedIOBase.detach)

Since 3.12 flushing has been necessary in GzipFile (see gh-105808 which was a release blocker), this makes that more visible. Users have been encountering as they upgrade to 3.12 (ex. gh-129726). 

There are a number of cases of unclosed file-like objects being deleted in CPython libraries and the test suite. This issue includes resolving those cases where the new ResourceWarning is emitted.

cc: @vstinner 

### Has this already been discussed elsewhere?

No response given

### Links to previous discussion of this feature:

https://github.com/python/cpython/issues/129726#issuecomment-2690022926

### Linked PRs
* gh-130905

**Comment by cmaloney** [comment]
Working on getting BufferedWriter warning to work, having a hard time as it involves adding a `tp_finalize`, and that interacts with the existing `tp_dealloc` which calls `_dealloc_warn` on itself and the underlying `raw` (often FileIO), which is how the FileIO close warning is emitted currently. I got it working well for `_pyio` but not sure how to get everything to fit for `_io.Buffered{Writer,Random}`. [What I currently have](https://github.com/python/cpython/compare/main...cmaloney:exp/writewarning?expand=1#diff-893faf8dc50d99bfa374e52717f4a7f6a2dd0020fbaa9dfc2cc45464b731d36a). At the moment struggling to figure out whats the right tweaks, particularly for tests which seem to have often evolved separately from the original need, and this changes behavior (ex. _pyio tests using _io by accident, checks for no warning and unraisable exceptions, etc). Wondering if with tp_finalize as it currently works could replace the existing _dealloc_warn / FileIO system with a simpler one...

While working on that, found the docs around `tp_finalize` pointed to the deprecated C APIs [PyErr_Fetch](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_Fetch) + [PyErr_Restore](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_Restore), made an issue + PR to update to [PyErr_GetRaisedException](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_GetRaisedException) and [PyErr_SetRaisedException](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_SetRaisedException): gh-131117

cc: @vstinner

**Comment by cmaloney** [comment]
Reducing the issue to just `GzipFile`, after a couple attempts at getting BufferedWriter working, have yet to find a reliable way. Is really neat how `_dealloc_warn` works with the `tp_finalize` in `IOBase`.
